src/ray/gcs/gcs_server/test/gcs_resource_report_poller_test.cc

```diff
@@ -0,0 +1,194 @@
+  RunPollingService();
+  ASSERT_TRUE(rpc_sent);
+}
+
+TEST_F(GcsResourceReportPollerTest, TestMaxInFlight) {
+  std::vector<std::shared_ptr<rpc::GcsNodeInfo>> nodes;
+  for (int i = 0; i < 200; i++) {
+    nodes.emplace_back(Mocker::GenNodeInfo());
+  }
+
+  std::vector<
+      std::function<void(const Status &, const rpc::RequestResourceReportReply &)>>
+      callbacks;
+
+  int num_rpcs_sent = 0;
+  request_report_ =
+      [&](const rpc::Address &, std::shared_ptr<rpc::NodeManagerClientPool> &,
+          std::function<void(const Status &, const rpc::RequestResourceReportReply &)>
+              callback) {
+        num_rpcs_sent++;
+        callbacks.emplace_back(callback);
+      };
+
+  for (const auto &node : nodes) {
+    gcs_resource_report_poller_.HandleNodeAdded(node);
+  }
+  RunPollingService();
+  ASSERT_EQ(num_rpcs_sent, 100);
+
+  for (int i = 0; i < 100; i++) {
+    const auto &callback = callbacks[i];
+    rpc::RequestResourceReportReply temp;
+    callback(Status::OK(), temp);
+  }
+  RunPollingService();
+  // Requests are sent as soon as possible (don't wait for the next tick).
+  ASSERT_EQ(num_rpcs_sent, 200);
+}
+
+TEST_F(GcsResourceReportPollerTest, TestPrioritizeNewNodes) {
+  std::vector<std::shared_ptr<rpc::GcsNodeInfo>> nodes;
+  for (int i = 0; i < 200; i++) {
+    nodes.emplace_back(Mocker::GenNodeInfo());
+  }
+
+  std::vector<
+      std::function<void(const Status &, const rpc::RequestResourceReportReply &)>>
+      callbacks;
+  std::unordered_set<NodeID> nodes_requested;
+
+  int num_rpcs_sent = 0;
+  request_report_ =
+      [&](const rpc::Address &address, std::shared_ptr<rpc::NodeManagerClientPool> &,
+          std::function<void(const Status &, const rpc::RequestResourceReportReply &)>
+              callback) {
+        num_rpcs_sent++;
+        callbacks.emplace_back(callback);
+        auto node_id = NodeID::FromBinary(address.raylet_id());
+        nodes_requested.insert(node_id);
+      };
+
+  for (int i = 0; i < 100; i++) {
+    const auto &node = nodes[i];
+    gcs_resource_report_poller_.HandleNodeAdded(node);
+  }
+  RunPollingService();
+  ASSERT_EQ(num_rpcs_sent, 100);
+
+  Tick(50);
+  RunPollingService();
+  ASSERT_EQ(num_rpcs_sent, 100);
+  ASSERT_EQ(nodes_requested.size(), 100);
+
+  for (int i = 0; i < 100; i++) {
+    const auto &callback = callbacks[i];
+    rpc::RequestResourceReportReply temp;
+    callback(Status::OK(), temp);
+  }
+  RunPollingService();
+  ASSERT_EQ(num_rpcs_sent, 100);
+  ASSERT_EQ(nodes_requested.size(), 100);
+
+  // The first wave of nodes should run at t=150.
+  Tick(50);
+  RunPollingService();
+  ASSERT_EQ(num_rpcs_sent, 100);
+  ASSERT_EQ(nodes_requested.size(), 100);
+
+  // We shouuld now request from the new nodes before repeating the first wave of nodes.
+  for (int i = 100; i < 200; i++) {
+    const auto &node = nodes[i];
+    gcs_resource_report_poller_.HandleNodeAdded(node);
+  }
+  RunPollingService();
+  ASSERT_EQ(num_rpcs_sent, 200);
```

Comment:
Shouldn't this check whether it was the new nodes that were requested?

Reply:
I think that was the intention of the second assert. If we send 200 requests and the set of unique nodes is 200, they must've been the new nodes right?